lib/src/set_multimap/set_multimap_builder.dart

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2015, Google Inc. Please see the AUTHORS file for details.
+// All rights reserved. Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+part of built_collection.set_multimap;
+
+/// The Built Collection builder for [BuiltSetMultimap].
+///
+/// It implements the mutating part of the [SetMultimap] interface.
+///
+/// See the
+/// [Built Collection library documentation](#built_collection/built_collection)
+/// for the general properties of Built Collections.
+class SetMultimapBuilder<K, V> {
+  Map<K, BuiltSet<V>> _builtMap;
+  BuiltSetMultimap<K, V> _builtMapOwner;
+  Map<K, SetBuilder<V>> _builderMap;
+
+  /// Instantiates with elements from a [Map], [SetMultimap] or
+  /// [BuiltSetMultimap].
+  ///
+  /// Must be called with a generic type parameter.
+  ///
+  /// Wrong: `new SetMultimapBuilder({1: ['1'], 2: ['2'], 3: ['3']})`.
+  ///
+  /// Right: `new SetMultimapBuilder<int, String>({1: ['1'], 2: ['2'], 3: ['3']})`,
+  ///
+  /// Rejects nulls. Rejects keys and values of the wrong type.
+  factory SetMultimapBuilder([multimap = const {}]) {
+    return new SetMultimapBuilder<K, V>._uninitialized()..replace(multimap);
+  }
+
+  /// Converts to a [BuiltSetMultimap].
+  ///
+  /// The `SetMultimapBuilder` can be modified again and used to create any number
+  /// of `BuiltSetMultimap`s.
+  BuiltSetMultimap<K, V> build() {
+    if (_builtMapOwner == null) {
+      for (final key in _builderMap.keys) {
+        final builtSet = _builderMap[key].build();
+        if (builtSet.isEmpty) {
+          _builtMap.remove(key);
+        } else {
+          _builtMap[key] = builtSet;
+        }
+      }
+
+      _builtMapOwner = new BuiltSetMultimap<K, V>._withSafeMap(_builtMap);
+    }
+    return _builtMapOwner;
+  }
+
+  /// Applies a function to `this`.
+  void update(updates(SetMultimapBuilder<K, V> builder)) {
+    updates(this);
+  }
+
+  /// Replaces all elements with elements from a [Map], [SetMultimap] or
+  /// [BuiltSetMultimap].
+  void replace(multimap) {
+    if (multimap is BuiltSetMultimap<K, V>) {
+      _setOwner(multimap);
+    } else if (multimap is Map ||
+        multimap is SetMultimap ||
+        multimap is BuiltSetMultimap) {
+      _setWithCopyAndCheck(multimap.keys, (k) => multimap[k]);
+    } else {
+      throw new ArgumentError(
+          'expected Map, SetMultimap or BuiltSetMultimap, got ${multimap.runtimeType}');
+    }
+  }
+
+  // Based on SetMultimap.
+
+  /// As [SetMultimap.add].
+  void add(K key, V value) {
+    _makeWriteableCopy();
+    _checkKey(key);
+    _checkValue(value);
+
+    final values = _builderMap[key];
+    if (values == null) {
+      var newValues = _builtMap[key];
+      if (newValues == null) {
+        newValues = new SetBuilder<V>();
+      } else {
+        newValues = newValues.toBuilder();
```

Comment:
I don't like this! newValues is even changing types between 83 and 85/87...

Instead, eg:
```
final values = _builderMap[key].putIfAbsent(key, () => _builtMap.containsKey(key) ? _builtMap[key].toBuilder() : new SetBuilder<V>());
values.add(value);
```

It would be easier if we allowed new SetBuilder(null) to work:
```final values = _builderMap.putIfAbsent(key, () => new SetBuilder<V>(_builtMap[key]));```